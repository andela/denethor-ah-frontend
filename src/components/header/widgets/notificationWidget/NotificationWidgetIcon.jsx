import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faBell } from '@fortawesome/free-regular-svg-icons';

const NotificationWidgetIcon = ({ toggleBubble, notifications }) => (
  <div className="header__notification-widget__icon">
    <button onClick={toggleBubble}>
      <FontAwesomeIcon
        className={
          `header__notification-widget__icon__bell${notifications.some(({ read }) => !read)
            ? ' unread'
            : ''}`}
        icon={faBell}
        size="2x"
        onClick={toggleBubble}
      />
    </button>
  </div>
);

NotificationWidgetIcon.propTypes = {
  toggleBubble: PropTypes.func,
  notifications: PropTypes.array
};

const mapStateToProps = ({ profile: { imageUrl } }) => ({ imageUrl });

export default connect(mapStateToProps)(NotificationWidgetIcon);
