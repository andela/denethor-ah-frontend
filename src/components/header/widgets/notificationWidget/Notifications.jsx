import React from 'react';
import uuid from 'uuid/v1';
import NotificationItem from './Notification';

const Notifications = (notifications) => () => (
  <div className="header__notification-widget__preview-pane">
    <ul>
    {
      notifications[0]
        ?
          notifications.map((notification) =>
            <NotificationItem key={uuid()} notification={notification} />
          )
        : 
          <li className="header__notification-widget__preview-pane__no-notifications">
            <p>No notifications</p>
          </li>
      }
    </ul>
  </div>
);

export default Notifications;
