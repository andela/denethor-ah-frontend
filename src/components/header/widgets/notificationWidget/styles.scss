.header__notification-widget {
  display: flex;
  align-items: center;
  .header__notification-widget__icon {
    button {
      background-color: transparent;
      border: none;
      outline: none;
      .header__notification-widget__icon__bell {
        color: #EBB03D;
        &.unread {
          animation: ring 4s .7s ease-in-out infinite, glowing 2s infinite,;
          transform-origin: 50% 4px;
        }
      }
    }
  }
  ul {
    li {
      a {
        display: flex;
        width: 100%;
        p {
          width: 100%;
          &.unread {
            font-family: Montserrat-SemiBold;
          }
        }
        small {
          color: #777;
          padding-left: 0.5rem;
        }
      }
    }
  }

  .header__widget__bubble__content {
    &:after {
      right: 4.7rem;
    }
  }
}

.notification-widget__toast {
  color: #333;
  position: fixed;
  top: 5rem;
  a {
    text-decoration: none;
    color: #333;
  }
}

@keyframes glowing {
  0% {
    color: darken(#EBB03D, 10%);
    filter: drop-shadow(0 0 -10px #fff); }
  40% {
    color: lighten(#EBB03D, 10%);
    filter: drop-shadow(0 0 20px lighten(#EBB03D, 20%));
  }
  60% {
    color: lighten(#EBB03D, 10%);
    filter: drop-shadow(0 0 20px lighten(#EBB03D, 20%));
  }
  100% {
    color: darken(#EBB03D, 10%);
    filter: drop-shadow(0 0 -10px #fff);
  }
}

@keyframes ring {
  0% { transform: rotate(0); }
  1% { transform: rotate(30deg); }
  3% { transform: rotate(-28deg); }
  5% { transform: rotate(34deg); }
  7% { transform: rotate(-32deg); }
  9% { transform: rotate(30deg); }
  11% { transform: rotate(-28deg); }
  13% { transform: rotate(26deg); }
  15% { transform: rotate(-24deg); }
  17% { transform: rotate(22deg); }
  19% { transform: rotate(-20deg); }
  21% { transform: rotate(18deg); }
  23% { transform: rotate(-16deg); }
  25% { transform: rotate(14deg); }
  27% { transform: rotate(-12deg); }
  29% { transform: rotate(10deg); }
  31% { transform: rotate(-8deg); }
  33% { transform: rotate(6deg); }
  35% { transform: rotate(-4deg); }
  37% { transform: rotate(2deg); }
  39% { transform: rotate(-1deg); }
  41% { transform: rotate(1deg); }

  43% { transform: rotate(0); }
  100% { transform: rotate(0); }
}
