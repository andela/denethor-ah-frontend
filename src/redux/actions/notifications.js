import socket from '../../utils/socket';
import { REMOVE_NOTIFICATION } from './types';

export const removeNotifications = (articleId) => async (dispatch, getState) => {
  dispatch({
    type: REMOVE_NOTIFICATION,
    payload: articleId
  });

  const { notifications } = getState();

  socket.emit('setRemainingNotifications', notifications);
}
