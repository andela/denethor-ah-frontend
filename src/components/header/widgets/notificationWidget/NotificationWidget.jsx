import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter, Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { toast } from 'react-toastify';
import Bubble from '../bubble';
import NotificationWidgetIcon from './NotificationWidgetIcon';
import Notifications from './Notifications';
import removeNotifications from '../../../../redux/actions/notifications';
import './styles.scss';


export class NotificationWidget extends Component {
  state = {
    showBubble: false,
    newNotifications: 0,
    toastId: 'notification'
  }

  componentDidMount () {
    const {
      props: { notifications = [], removeNotifications },
      state: { toastId }
    } = this;
  
    const newNotification = notifications.find(({ read }) => !read);
    const newNotificationsInProps = notifications.reduce((count, { read }) => count + (read ? 0 : 1), 0);

    if(newNotification) {
      const { message, articleId } = newNotification;

      if (!toast.isActive(toastId)) {
        const Notification = ({ closeToast }) => (
          <Link
          to={`/articles/${articleId}`}
          onClick={() => {
            removeNotifications(articleId);
            closeToast()
          }}>
            {message}
          </Link>
        );

        toast(<Notification />, {
          toastId,
          autoClose: 10000,
          hideProgressBar: true,
          className: 'notification-widget__toast',
          closeOnClick: false
        });
      }
      const newNotifications = newNotificationsInProps;

      this.setState({ newNotifications });
    }
  }

  componentDidUpdate () {
    const {
      props: { notifications = [], removeNotifications },
      state: { newNotifications: newNotificationsInState, toastId }
    } = this;

    const newNotificationsInProps = notifications.reduce((count, { read }) => count + (read ? 0 : 1), 0);

    if(newNotificationsInProps > newNotificationsInState) {
      const { message, articleId } = notifications.find(({ read }) => !read);

      if (!toast.isActive(toastId)) {
        const Notification = ({ closeToast }) => (
          <Link
          to={`/articles/${articleId}`}
          onClick={() => {
            removeNotifications(articleId);
            closeToast()
          }}>
            {message}
          </Link>
        );

        toast(<Notification />, {
          toastId,
          autoClose: 10000,
          hideProgressBar: true,
          className: 'notification-widget__toast',
          closeOnClick: false
        });
      }
      const newNotifications = newNotificationsInProps;

      this.setState({ newNotifications });
    }
  }

  toggleBubble = (showBubble) => {
    if(showBubble !== undefined) {
      return this.setState({ showBubble });
    }
    this.setState(({ showBubble }) => ({ showBubble: !showBubble }));
  }

  render() {
    const {
      toggleBubble,
      props: { notifications = [] },
      state: { showBubble }
    } = this;

    return (
      <div className="header__notification-widget">
        <NotificationWidgetIcon notifications={notifications} toggleBubble={toggleBubble} />
        {showBubble && <Bubble BubbleContent={Notifications(notifications)} toggleBubble={toggleBubble}/>}
      </div>
    );
  }
}

NotificationWidget.propTypes = {
  history: PropTypes.object,
  notifications: PropTypes.array,
  removeNotifications: PropTypes.func
}

const mapStateToProps = ({ notifications }) => ({ notifications });

const mapDispatchToProps = (dispatch) => ({
  removeNotifications: (articleId) => dispatch(removeNotifications(articleId)),
});

export default (connect(mapStateToProps, mapDispatchToProps)(withRouter(NotificationWidget)));
