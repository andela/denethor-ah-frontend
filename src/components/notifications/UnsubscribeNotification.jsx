import React, { Component } from 'react';
import axios from 'axios';
import { toast} from 'react-toastify';
import PropTypes from 'prop-types';
import '../redirect/styles.scss';

export class  UnsubscribeNotification extends Component {
  async componentDidMount() {
    try {
      const link = `${process.env.API_ROOT_URL}/users/${this.props.match.params.id}/unsubscribe`;
      const { data: { message } } = await axios.patch(link);
      this.props.history.push('/');
      toast.success(message, { className: 'toast-custom-style' });
    } catch({ response }) {
          toast.error('Unknown error. Try again',  { className: 'toast-custom-style' });
         this.props.history.push('/signup');
      }
  }
  render() {
    return (
      <div className="redirect-page">
        <div className="spinner-container-redirect-page">
          <div className="spinner spinner--redirect-page"></div>
        </div>
      </div>
    );
  }
}
UnsubscribeNotification.propTypes = { 
  history: PropTypes.object,
  match: PropTypes.object
}
export default UnsubscribeNotification;