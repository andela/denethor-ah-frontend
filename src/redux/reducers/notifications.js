import { SET_NOTIFICATIONS, REMOVE_NOTIFICATION } from '../actions/types';

const notificationsDefaultState = [];

const notificationsReducer = (state = notificationsDefaultState, action) => {
  switch (action.type) {
    case SET_NOTIFICATIONS:
      return action.payload;

    case REMOVE_NOTIFICATION:
      return state.map((notification) => ({
        ...notification,
        ...(notification.articleId === action.payload && { read: true })
      }));

    default:
      return state;
  }
};

export default notificationsReducer;
