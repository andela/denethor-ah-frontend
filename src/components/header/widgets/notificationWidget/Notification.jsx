import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import moment from 'moment';
import { removeNotifications } from '../../../../redux/actions/notifications';

moment.updateLocale('en', {
  relativeTime : {
      future: "in %s",
      past:   "%s",
      s  : '%ds',
      ss : '%ds',
      m:  "%dm",
      mm: "%dm",
      h:  "%dh",
      hh: "%dh",
      d:  "%dd",
      dd: "%dd",
      M:  "a month",
      MM: "%d months",
      y:  "a year",
      yy: "%d years"
  }
});


export const Notification = props => {
  const { message, articleId, time, read } = props.notification;
  
  return (
    <li>
      <Link to={`/articles/${articleId}`} onClick={() => props.removeNotifications(articleId)}>
        <p className={read ? '' : 'unread'}>{message}</p>
        <span><small>{moment(time).fromNow()}</small></span>
      </Link>
    </li>
  );
}

Notification.propTypes = {
  notification: PropTypes.object,
  removeNotifications: PropTypes.func
}

export const mapDispatchToProps = (dispatch) => ({
  removeNotifications: (articleId) => dispatch(removeNotifications(articleId))
});

export default connect(undefined, mapDispatchToProps)(Notification);
