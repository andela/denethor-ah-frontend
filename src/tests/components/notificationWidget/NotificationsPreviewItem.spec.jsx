import React from 'react';
import { shallow, mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import { Notification, mapDispatchToProps } from '../../../components/header/widgets/notificationWidget/Notification';
import notifications from './mock-notifications';
import { removeNotifications } from '../../../redux/actions/notifications';

jest.mock('../../../utils/socket');
jest.mock('../../../redux/actions/notifications');

test('Should render static component', () => {
  const props = {
    notification: { ...notifications[0], time: 12345 },
    removeNotifications: () => {}
  }
  const wrapper = shallow(<Notification { ...props } />);

  expect(wrapper).toMatchSnapshot();
});

test('Should map dispatch to props', () => {
  const dispatch = jest.fn();
  removeNotifications.mockReturnValue('notification removed');

  const { removeNotifications: stateRemoveNotifications } = mapDispatchToProps(dispatch);

  stateRemoveNotifications(3);

  expect(dispatch).toHaveBeenLastCalledWith('notification removed');
});


test('Should remvove article on click', () => {
  const removeNotifications = jest.fn();
  const props = {
    notification: { ...notifications[0], time: 12345 },
    removeNotifications,
  }
  const wrapper = mount(<BrowserRouter><Notification { ...props } /></BrowserRouter>);

  wrapper.find('a').simulate('click');

  expect(removeNotifications).toHaveBeenLastCalledWith(notifications[0].articleId);
});
