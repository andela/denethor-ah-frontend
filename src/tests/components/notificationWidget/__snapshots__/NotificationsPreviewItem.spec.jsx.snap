// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render static component 1`] = `
<li>
  <Link
    onClick={[Function]}
    replace={false}
    to="/articles/1"
  >
    <p
      className="unread"
    >
      this is the first notification
    </p>
    <span>
      <small>
        49 years
      </small>
    </span>
  </Link>
</li>
`;
